src/cascadia/TerminalConnection/ConptyConnection.cpp

```diff
@@ -34,223 +158,176 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
-                    *phInput = inPipeOurSide;
-                    *phOutput = outPipeOurSide;
-                }
-                CloseHandle(outPipePseudoConsoleSide);
-            }
-            else
-            {
-                CloseHandle(inPipeOurSide);
-            }
-            CloseHandle(inPipePseudoConsoleSide);
+            ResizePseudoConsole(_hPC.get(), { Utils::ClampToShortMax(columns, 1), Utils::ClampToShortMax(rows, 1) });
         }
-        return hr;
     }
 
-    // Prepares the `lpAttributeList` member of a STARTUPINFOEX for attaching a
-    //      client application to a pseudoconsole.
-    // Prior to calling this function, hPty should be initialized with a call to
-    //      CreatePseudoConsole, and the pAttrList should be initialized with a call
-    //      to InitializeProcThreadAttributeList. The caller of
-    //      InitializeProcThreadAttributeList should add one to the dwAttributeCount
-    //      param when creating the attribute list for usage by this function.
-    HRESULT _AttachPseudoConsole(HPCON hPC, LPPROC_THREAD_ATTRIBUTE_LIST lpAttributeList)
+    void ConptyConnection::Close()
     {
-        BOOL fSuccess = UpdateProcThreadAttribute(lpAttributeList,
-                                                  0,
-                                                  PROC_THREAD_ATTRIBUTE_PSEUDOCONSOLE,
-                                                  hPC,
-                                                  sizeof(HPCON),
-                                                  NULL,
-                                                  NULL);
-        return fSuccess ? S_OK : HRESULT_FROM_WIN32(GetLastError());
-    }
+        if (!_connected)
+        {
+            return;
+        }
 
-    void ConptyConnection::_CreatePseudoConsole()
-    {
-        bool fSuccess;
+        if (!_closing.exchange(true))
+        {
+            _clientExitWait.reset(); // immediately stop waiting for the client to exit.
 
-        COORD dimensions{ _initialCols, _initialRows };
-        THROW_IF_FAILED(_CreatePseudoConsoleAndHandles(dimensions, 0, &_inPipe, &_outPipe, &_hPC));
+            _hPC.reset();
 
-        STARTUPINFOEX siEx;
-        siEx = { 0 };
-        siEx.StartupInfo.cb = sizeof(STARTUPINFOEX);
-        size_t size;
-        InitializeProcThreadAttributeList(NULL, 1, 0, (PSIZE_T)&size);
-        BYTE* attrList = new BYTE[size];
-        siEx.lpAttributeList = reinterpret_cast<PPROC_THREAD_ATTRIBUTE_LIST>(attrList);
-        fSuccess = InitializeProcThreadAttributeList(siEx.lpAttributeList, 1, 0, (PSIZE_T)&size);
-        THROW_LAST_ERROR_IF(!fSuccess);
+            _inPipe.reset();
+            _outPipe.reset();
 
-        THROW_IF_FAILED(_AttachPseudoConsole(_hPC, siEx.lpAttributeList));
+            // Tear down our output thread -- now that the output pipe was closed on the
+            // far side, we can run down our local reader.
+            WaitForSingleObject(_hOutputThread.get(), INFINITE);
```

Comment:
Do we need to check the errors off of WaitForSingleObject? (and below)

Reply:
I don't believe so, but I can log them